port/osx/omrosdump.c

```diff
@@ -0,0 +1,437 @@
+static int corefile_fd = -1;
+
+static kern_return_t coredump_to_file(mach_port_t, pid_t);
+static kern_return_t list_thread_commands(mach_port_t, struct thread_command_full_64 **, natural_t *);
+static kern_return_t list_segment_commands(mach_port_t, struct segment_command_64 **, natural_t *);
+
+static kern_return_t coredump_to_file(mach_port_t task_port, pid_t pid)
+{
+	kern_return_t kr = KERN_SUCCESS;
+	struct mach_header_64 mh64;
+	struct segment_command_64 *segments = NULL;
+	struct thread_command_full_64 *threads = NULL;
+	natural_t segment_count;
+	natural_t thread_count;
+	natural_t cpu_count;
+	processor_basic_info_t proc_info_array;
+	mach_msg_type_number_t info_count;
+	off_t file_off = 0;
+	uint64_t seg_file_off = 0;
+	ssize_t written = 0;
+
+	corefile_fd = open(corefile_name, O_RDWR | O_CREAT | O_EXCL, 0600);
+
+	if (-1 == corefile_fd) {
+		perror("open()");
+		goto done;
+	}
+
+	kr = host_processor_info(mach_host_self(), PROCESSOR_BASIC_INFO, &cpu_count,
+		(processor_info_array_t *)&proc_info_array, &info_count);
+	if (KERN_SUCCESS != kr) {
+		mach_error("failed to get processor info:\n", kr);
+		goto done;
+	}
+
+	mh64.magic = MH_MAGIC_64;
+	/* host_processor_info does not return a value with the 64-bit flag set so
+	 * set manually since we only deal with 64-bit on OSX
+	 */
+	mh64.cputype = proc_info_array[0].cpu_type | CPU_ARCH_ABI64;
+	mh64.cpusubtype = proc_info_array[0].cpu_subtype;
+	/* update ncmds and sizeofcmds after processing thread and segment commands */
+	mh64.ncmds = 0;
+	mh64.sizeofcmds = 32;
+	mh64.filetype = MH_CORE;
+
+	file_off += sizeof(struct mach_header_64);
+	segments = NULL;
+	kr = list_thread_commands(task_port, &threads, &thread_count);
+	if (KERN_SUCCESS != kr) {
+		mach_error("error getting thread command data:\n", kr);
+		goto done;
+	}
+
+	for (natural_t i = 0; i < thread_count; i++) {
+		/* write each portion separately due to structure padding */
+		written = pwrite(corefile_fd, &threads[i].cmd, 4, file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		written = pwrite(corefile_fd, &threads[i].cmdsize, 4, file_off + 4);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+
+		file_off += 8;
+		written = pwrite(corefile_fd, &threads[i].thread_state, sizeof(x86_thread_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(x86_thread_state_t);
+		written = pwrite(corefile_fd, &threads[i].float_state, sizeof(x86_float_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(x86_float_state_t);
+		written = pwrite(corefile_fd, &threads[i].exceptions, sizeof(x86_exception_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing thread commands:");
```

Comment:
Why the different message here ('commands' could be removed or it should match the data: 'exceptions')?